// locutus/php/array/index.js
'use strict';

module.exports['array_change_key_case'] = require('./array_change_key_case');
module.exports['array_chunk'] = require('./array_chunk');
module.exports['array_combine'] = require('./array_combine');
module.exports['array_count_values'] = require('./array_count_values');
module.exports['array_diff'] = require('./array_diff');
module.exports['array_diff_assoc'] = require('./array_diff_assoc');
module.exports['array_diff_key'] = require('./array_diff_key');
module.exports['array_diff_uassoc'] = require('./array_diff_uassoc');
module.exports['array_diff_ukey'] = require('./array_diff_ukey');
module.exports['array_fill'] = require('./array_fill');
module.exports['array_fill_keys'] = require('./array_fill_keys');
module.exports['array_filter'] = require('./array_filter');
module.exports['array_flip'] = require('./array_flip');
module.exports['array_intersect'] = require('./array_intersect');
module.exports['array_intersect_assoc'] = require('./array_intersect_assoc');
module.exports['array_intersect_key'] = require('./array_intersect_key');
module.exports['array_intersect_uassoc'] = require('./array_intersect_uassoc');
module.exports['array_intersect_ukey'] = require('./array_intersect_ukey');
module.exports['array_key_exists'] = require('./array_key_exists');
module.exports['array_keys'] = require('./array_keys');
module.exports['array_map'] = require('./array_map');
module.exports['array_merge'] = require('./array_merge');
module.exports['array_merge_recursive'] = require('./array_merge_recursive');
module.exports['array_multisort'] = require('./array_multisort');
module.exports['array_pad'] = require('./array_pad');
module.exports['array_pop'] = require('./array_pop');
module.exports['array_product'] = require('./array_product');
module.exports['array_push'] = require('./array_push');
module.exports['array_rand'] = require('./array_rand');
module.exports['array_reduce'] = require('./array_reduce');
module.exports['array_replace'] = require('./array_replace');
module.exports['array_replace_recursive'] = require('./array_replace_recursive');
module.exports['array_reverse'] = require('./array_reverse');
module.exports['array_search'] = require('./array_search');
module.exports['array_shift'] = require('./array_shift');
module.exports['array_slice'] = require('./array_slice');
module.exports['array_splice'] = require('./array_splice');
module.exports['array_sum'] = require('./array_sum');
module.exports['array_udiff'] = require('./array_udiff');
module.exports['array_udiff_assoc'] = require('./array_udiff_assoc');
module.exports['array_udiff_uassoc'] = require('./array_udiff_uassoc');
module.exports['array_uintersect'] = require('./array_uintersect');
module.exports['array_uintersect_uassoc'] = require('./array_uintersect_uassoc');
module.exports['array_unique'] = require('./array_unique');
module.exports['array_unshift'] = require('./array_unshift');
module.exports['array_values'] = require('./array_values');
module.exports['array_walk'] = require('./array_walk');
module.exports['array_walk_recursive'] = require('./array_walk_recursive');
module.exports['arsort'] = require('./arsort');
module.exports['asort'] = require('./asort');
module.exports['count'] = require('./count');
module.exports['current'] = require('./current');
module.exports['each'] = require('./each');
module.exports['end'] = require('./end');
module.exports['in_array'] = require('./in_array');
module.exports['key'] = require('./key');
module.exports['krsort'] = require('./krsort');
module.exports['ksort'] = require('./ksort');
module.exports['natcasesort'] = require('./natcasesort');
module.exports['natsort'] = require('./natsort');
module.exports['next'] = require('./next');
module.exports['pos'] = require('./pos');
module.exports['prev'] = require('./prev');
module.exports['range'] = require('./range');
module.exports['reset'] = require('./reset');
module.exports['rsort'] = require('./rsort');
module.exports['shuffle'] = require('./shuffle');
module.exports['sizeof'] = require('./sizeof');
module.exports['sort'] = require('./sort');
module.exports['uasort'] = require('./uasort');
module.exports['uksort'] = require('./uksort');
module.exports['usort'] = require('./usort');
