// ember-cli-lodash-subset/lodash.subset.js
import clone from 'lodash-es/clone';
import cloneDeep from 'lodash-es/cloneDeep';
import compact from 'lodash-es/compact';
import debounce from 'lodash-es/debounce';
import defaults from 'lodash-es/defaults';
import defaultsDeep from 'lodash-es/defaultsDeep';
import filter from 'lodash-es/filter';
import find from 'lodash-es/find';
import forEach from 'lodash-es/forEach';
import forOwn from 'lodash-es/forOwn';
import includes from 'lodash-es/includes';
import intersection from 'lodash-es/intersection';
import isEqual from 'lodash-es/isEqual';
import isNull from 'lodash-es/isNull';
import map from 'lodash-es/map';
import merge from 'lodash-es/merge';
import omitBy from 'lodash-es/omitBy';
import reject from 'lodash-es/reject';
import remove from 'lodash-es/remove';
import union from 'lodash-es/union';
import uniq from 'lodash-es/uniq';
import uniqBy from 'lodash-es/uniqBy';
import values from 'lodash-es/values';
import zipObject from 'lodash-es/zipObject';

module.exports = {
  clone: clone,
  cloneDeep: cloneDeep,
  compact: compact,
  debounce: debounce,
  defaults: defaults,
  defaultsDeep: defaultsDeep,
  filter: filter,
  find: find,
  forEach: forEach,
  forOwn: forOwn,
  includes: includes,
  intersection: intersection,
  isEqual: isEqual,
  isNull: isNull,
  map: map,
  merge: merge,
  omitBy: omitBy,
  reject: reject,
  remove: remove,
  union: union,
  uniq: uniq,
  uniqBy: uniqBy,
  values: values,
  zipObject: zipObject,
};
