// nav-frontend-paneler/src/index.tsx
import * as PT from 'prop-types';
import * as React from 'react';
import * as classNames from 'classnames';
import 'nav-frontend-paneler-style';

const cls = (className, border) => classNames('panel', className, {
    'panel--border': border
});

export interface PanelBaseProps {
    /**
     * Innhold
     */
    children: React.ReactNode | React.ReactChild | React.ReactChildren;
    /**
     * Egendefinert klassenavn
     */
    className?: string;
    /**
     * Hvis komponenten skal brukes på hvit bakgrunn kan denne brukes for å gi den en border
     */
    border?: boolean;
}

class PanelBase extends React.Component<PanelBaseProps> {
    static defaultProps: Partial<PanelBaseProps> = {
        border: false
    };
    render() {
        const { children, className, border, ...props } = this.props;
        return (
            <div className={cls(className, border)} {...props}>{children}</div>
        );
    }
}

(PanelBase as React.ComponentClass).propTypes = {
    className: PT.string,
    children: PT.node.isRequired,
    border: PT.bool
};
(PanelBase as React.ComponentClass).defaultProps = {
    className: undefined,
    border: false
};

export default PanelBase;

export const Panel = (props) => <PanelBase children={props.children} {...props}/>;
